packages/angular_devkit/core/src/jobs/README.md

```diff
@@ -0,0 +1,319 @@
+
+```typescript
+import { Observable } from 'rxjs';
+import { jobs } from '@angular-devkit/core';
+
+const input = {
+  type: 'array', items: { type: 'number' },
+};
+const output = {
+  type: 'number',
+};
+
+export function add(input: number[]): Observable<jobs.JobEvent<number>> {
+  return new Observable(o => {
+    o.next({ kind: jobs.JobEventKind.Start });
+    o.next({
+      kind: jobs.JobEventKind.Output,
+      output: input.reduce((total, curr) => total + curr, 0),
+    });
+    o.next({ kind: jobs.JobEventKind.End });
+  });
+}
+
+// Add a property to `add` to make it officially a JobHandler. The Job system does not recognize any
+// function as a JobHandler.
+add.jobDescription = {
+  name: 'add',
+  input: input,
+  output: output,
+};
+
+// Calling the job, and logging its output.
+declare const scheduler: jobs.Scheduler;
+scheduler.schedule('add', [1, 2, 3, 4])
+    .output.subscribe(x => console.log(x));  // Will output 10.
+```
+
+This is a lot of boilerplate, so we made some helpers to improve readability and manage input 
+and output automatically;
+
+```typescript
+// Add is a JobHandler function, like the above.
+export const add = jobs.createJob<number[], number>(
+  input => input.reduce((total, curr) => total + curr, 0),
+);
+
+// Schedule like above.
+```
+
+You can also return a Promise or an Observable, as jobs are asynchronous. This helper will set 
+progress to 0, start and end events appropriately, and will pass in a logger. It will also manage
+channels automatically (see below).
+
+A more complex job can be declared like this:
+
+```typescript
+import { Observable } from 'rxjs';
+import { JobRegistry } from './api'
+
+// Show progress with each count. Output "more" in a channel.
+export const count = jobs.createJob<number, number>(
+  // Receive a context that contains additional methods, like progress, logger and channels.
+  (input: number, { progress, channels }) => new Observable<number>(o => {
+    let i = 0;
+    function doCount() {
+      o.next(i++);
+      progress(i / input);
+      channels['side'].next('more');
+  
+      if (i < input) {
+        setTimeout(doCount, 100);
+      } else {
+        o.complete();
+      }
+    }
+    setTimeout(doCount, 100);
+  }),
+  {
+      input: { type: 'number' },
+      output: { type: 'number' },
+      channels: {
+        'side': { type: 'string', const: 'more' },
+      },
+  },
+);
+
+// Get a hold of a scheduler that refers to the job above.
+declare const scheduler: jobs.Scheduler;
+
+const job = scheduler.schedule('count');
+job.channels['side'] && job.channels['side'].subscribe(x => console.log(x));
+const c = job.getChannel<string>('side', { type: 'string' });
+if (c) {
+  c.subscribe((x: string) => console.log(x));
+}
+```
+
+## Job Extension
+A Job can extend other jobs, meaning their inputs and outputs have to be a superset of the job 
+extended. This is useful for making compatible jobs that can be replaced by others.
```

Comment:
I think this section really needs some examples and more details. Things like a job extending each other, what it means to replace a job, and a replacement example.

Another thing that I don't understand is the usage pattern for jobs that have a superset of the input and outputs that is not. 

Consider I have a JobFoo, with known inputs and outputs. JobBar extends JobFoo, and has additional inputs and outputs.

If I want to transparently replace JobFoo with JobBar, I can only rely on using the known input/output for JobFoo. Thus the additional input/output that JobBar has will always use defaults.

I'd like to understand how this impacts the use cases that Schematics and Architect have for Jobs.

For instance, having a generic `node_install` job that is ultimately replaced by `npm_install` or `yarn_install`, or some other. 

Should the base `node_install` have only the intersection of the input/output of the (known) possible replacements? Or should it have some input/output that we expect to be ignored by some of the replacements?

Replacements are one of the key requirements of Jobs so it's important that this be clear.